olp-cpp-sdk-core/include/olp/core/thread/ExecutionContext.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2022 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <utility>
+
+#include <olp/core/client/ApiError.h>
+#include <olp/core/client/CancellationContext.h>
+
+namespace olp {
+namespace thread {
+
+using FailedCallback = std::function<void(client::ApiError)>;
+using CancelFuncType = std::function<void()>;
+
+/// Handles the cancellation, final, and failed mechanisms.
+class CORE_API ExecutionContext {
+  using ExecuteFuncType = std::function<client::CancellationToken()>;
+
+ public:
+  /// A default contructor, initializes the `ExecutionContextImpl` instance.
+  ExecutionContext();
+
+  /**
+   * @brief Gets the error information.
+   *
+   * @return The `ApiError` instance containing the error information.
+   */
+  const client::ApiError GetError() const;
```

Comment:
Making it const means we need to copy it and cannot move it, although there is a std::string included. You have two options based on how this is used:
- Make it `const ApiError&` means you can access it and if you need to keep it, copy it
- [Prefered] Make it `ApiError&&` and move it, as it seems to me the ExecutionContext is trown away either way after this.

Also, what happens if the user calls this on an `ExecutionContext` which is not failed yet? Maybe warn the user that he should check with `Failed()` before to calling this method.